core/src/main/java/org/apache/iceberg/BaseEntriesTable.java

```diff
@@ -125,31 +130,120 @@ ManifestFile manifest() {
+
+        return CloseableIterable.transform(
+            entries(requiredFileProjection),
+            entry -> withReadableMetrics(structProjection, entry, readableMetricsField));
       }
+    }
+
+    /**
+     * Remove virtual columns from the file projection and ensure that the underlying metrics used
+     * to create those columns are part of the file projection
+     *
+     * @return file projection with required columns to read readable metrics
+     */
+    private Schema requiredFileProjection() {
+      Schema projectionForReadableMetrics =
+          new Schema(
+              MetricsUtil.READABLE_METRIC_COLS.stream()
+                  .map(MetricsUtil.ReadableMetricColDefinition::originalCol)
+                  .collect(Collectors.toList()));
+      return TypeUtil.join(fileProjection, projectionForReadableMetrics);
+    }
 
-      // Project non-readable fields
-      Schema readSchema = ManifestEntry.wrapFileSchema(fileSchema.asStruct());
-      StructProjection projection = StructProjection.create(readSchema, schema);
-      return CloseableIterable.transform(prunedRows, projection::wrap);
+    private Schema removeReadableMetrics(Types.NestedField readableMetricsField) {
+      Set<Integer> readableMetricsIds = TypeUtil.getProjectedIds(readableMetricsField.type());
+      return TypeUtil.selectNot(projection, readableMetricsIds);
+    }
+
+    private StructProjection projectNonReadable(Schema projectedSchema) {
+      Schema manifestEntrySchema = ManifestEntry.wrapFileSchema(fileProjection.asStruct());
+      return StructProjection.create(manifestEntrySchema, projectedSchema);
+    }
+
+    private CloseableIterable<? extends ManifestEntry<? extends ContentFile<?>>> entries(
+        Schema newFileProjection) {
+      return ManifestFiles.open(manifest, io, specsById).project(newFileProjection).entries();
+    }
+
+    private StructLike withReadableMetrics(
+        StructProjection structProjection,
+        ManifestEntry<? extends ContentFile<?>> entry,
+        Types.NestedField readableMetricsField) {
+      int projectionColumnCount = projection.columns().size();
+      int metricsPosition = projection.columns().indexOf(readableMetricsField);
+
+      StructProjection entryStruct = structProjection.wrap((StructLike) entry);
+
+      StructType projectedMetricType =
+          projection.findField(MetricsUtil.READABLE_METRICS).type().asStructType();
+      MetricsUtil.ReadableMetricsStruct readableMetrics =
+          MetricsUtil.readableMetricsStruct(dataTableSchema, entry.file(), projectedMetricType);
+
+      return new ManifestEntryStructWithMetrics(
+          projectionColumnCount, metricsPosition, entryStruct, readableMetrics);
     }
 
     @Override
     public Iterable<FileScanTask> split(long splitSize) {
       return ImmutableList.of(this); // don't split
     }
   }
+
+  static class ManifestEntryStructWithMetrics implements StructLike {
```

Comment:
For another pr, I think we can combine this and BaseFilesTable's similar class (StructProjection is instance of StructLike)

Reply:
yeah that's a good idea. I think we can extract the common into separate StructLikeWithMetrics class to be used in both baseEntries and baseFiles